core/store/migrate/migrations/0254_log_poller_primary_keys.sql

```diff
@@ -0,0 +1,24 @@
+-- +goose Up
+
+ALTER TABLE evm.logs DROP CONSTRAINT logs_pkey;
+ALTER TABLE evm.logs ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
+CREATE UNIQUE INDEX idx_logs_chain_block_logindex ON evm.logs (evm_chain_id, block_number, log_index);
+ALTER TABLE evm.log_poller_blocks DROP CONSTRAINT log_poller_blocks_pkey;
+ALTER TABLE evm.log_poller_blocks ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_order_by_block;
+DROP INDEX IF EXISTS evm.idx_evm_log_poller_blocks_block_number_evm_chain_id;
+CREATE UNIQUE INDEX idx_log_poller_blocks_chain_block ON evm.log_poller_blocks (evm_chain_id, block_number DESC);
+DROP INDEX IF EXISTS evm.idx_evm_logs_ordered_by_block_and_created_at;
+CREATE INDEX idx_logs_chain_address_event_block_logindex ON evm.logs (evm_chain_id, address, event_sig, block_number, log_index);
```

Comment:
I think most of the queries of the logs table should be able to use it. For example, the query for `SelectLogs()` is:

```
    query := logsQuery(`
        WHERE evm_chain_id = :evm_chain_id
        AND address = :address
        AND event_sig = :event_sig
        AND block_number >= :start_block
        AND block_number <= :end_block
        ORDER BY block_number, log_index`)
```

Like most of the logs queries, it filtesr on chain id, address, and event sig and then sorts on block_number, log_index (the order in which logs appear on chain).

Reply:
I got rid of a similar index we had for `created_at`, which was used at one point for ordering instead of `block_number, log_index` but we no longer use `created_at` for anything since it doesn't reliably indicate the order in which the events occurred on chain.

Reply:
Ahh, right, makes sense

Reply:
Yeah, that makes sense, but isn't that already covered by one of the existing indexes?
```sql
create index evm_logs_idx
    on evm.logs (evm_chain_id, block_number, address, event_sig);

create index evm_logs_by_timestamp
    on evm.logs (evm_chain_id, address, event_sig, block_timestamp, block_number);
```

Usually, there are not many matching logs within a single block, so probably gain from adding log_index to the index won't be that high, right? Postgres should be able to sort those small datasets efficiently by using the log_index in memory.

What I'm trying to understand if it's really required by the changes from this PR. In the past, there were some teams (Automation?) very unhappy with us increasing the db size by adding new indexes. If this index is already covered by other ones then maybe we don't need it at this stage

Reply:
> postgres should be able to sort those small datasets efficiently by using the log_index in memory.

That's true, I'd be fine with omitting the `log_index` column as it's not adding much.  On the other hand, I wouldn't expect that including it would increase the size of that index much.

> isn't that already covered by one of the existing indexes?

I figured it probably can't use either of those for this, but admittedly I haven't tested this.

My general understanding of how these BTREE indices work based on what I've seen in previous plans is that you have to have all of the columns you're filtering on first, and then the sort columns after.  Assuming that is how BTREE's work, the `evm_chain_id, block_number, address, event_sig` index will help for block range queries where you're looking for events of specific types within a particular block range.  While this one would help with ordering the results for events of a specific type.

The one with `block_timestamp` in it seems more likely to work. If it had a way of knowing for sure that block numbers and block timestamps will be ordered in the same way, then it should be able to use it... but I don't think it can know that so my assumption would be that it  probably won't use it.

Reply:
Side thought:  from a database perspective, `block_timestamp` really ought to have been a column of the `blocks` table rather than the `logs` table. We could have easily joined those on block_number to search by block_timestamp without much cost. I guess the reason why it's in the `logs` table is mostly so that it can be returned by the queries that return `[]Log` and to keep the `logs` schema in line with that data struct. Not sure if this decision might be something to revisit in the future.

Reply:
> the evm_chain_id, block_number, address, event_sig index will help for block range queries where you're looking for events of specific types within a particular block range

After writing this I realized this is not right. If this were the only use case, then we'd be able to get rid of the  `evm_chain_id, block_number, address, event_sig` index, and just use `evm_chain_id, address, event_sig, block_number` for everything... because the SELECT queries where we're looking for events in a particular block range also filter on event_sig and address... so it can just use the `evm_chain_id, address, event_sig, block_number, log_index` index.

Unfortunately, I just looked through the code and I do see one query that filters on block range but not address or event sig. It's where we delete all of the logs and blocks after a particular block due to a re-org:
```
DELETE FROM evm.logs
                         WHERE evm_chain_id = $1
                          AND block_number >= $2
                          AND block_number <= (SELECT MAX(block_number) FROM evm.logs WHERE evm_chain_id = $1)
 ```
So I think we can get rid of the `address, event_sig` columns at the end of this index to save some space.